### ISSUE TITLE
live setpoint does not work

### ISSUE THREAD

**Issue by shengb** [frib-high-level-controls/save-set-restore#16]
Actually, this might be from the conceptual difference.
MASAR trades all pvs same, do not has logic of set point PV or read back PV.
2 issues:
1. The "Live Setpoint" does not get the real value from IOC although the IOC is running.
2. Since the saved value is called "Stored Value", another column should be "Live Value" instead of "Live Setpoint".

**Comment by jbobnar** [comment]
In ideal world there exist a service, which provides a readback PV for "every" setpoint PV. The idea is that (some day) this will be channel finder. However, populating the channel finder with this info today is not feasible, therefore a decision was made to fallback to storing the readback pvs (and threshold values) into the set file. There is still an extension point, through which you will be able to provide the readback names obtained the channel finder, when implemented. In such case it is irrelevant whether one is working with masar service, git or any other data provider. 

1. Please, explain under what condition this happens. However, you should note that CSS connects to PVs directly, while python masar client queries the service for the live values. If your IOC is behind a firewall, that might be the cause of the problem.
2. I wanted to make a clear separation between Live Readback and Live Setpoint. If 'Live Value' is generally preferred, I can of course change that.

**Comment by berryma4** [comment]
I prefer the file at this time (not channelfinder), not only because of the above, but we would need to see the benefits to the extra complexity and dependency.  And, In the end I see channelfinder, or other service, populating the file.  So, the file would still exist.  This is because it is a "snapshot", and we shouldn't rely on mutable data for a record.

I prefer Live Readback and Live Setpoint, as this is what was preferred from our meetings (including A1900, ReA, and ECR representatives).

**Comment by shengb** [comment]
Here we are talking about 2 different topics:
1. Shall we always pair setpoint & readback? It is a general topic, and always has different opinions.
2. PV connections.

Let's focus on the issues 2 first.
I was trying it on my Mac laptop. Everything (MASAR service, testing IOC, and CS-Studio) runs on the same machine.
I am able to do a caget from terminal, and firewall is off.
But the Live value is empty.

It is OK on a linux machine.

**Comment by berryma4** [comment]
Careful, cs-studio/diirt holds their own settings for epics in
/etc/cs-studio/diirt (not the same as commandline caget).
Our default for EPICS_CA_ADDR_LIST is to go through the gateway.  You may
need to add localhost, if you are doing localhost work.

**Comment by shengb** [comment]
Thanks for letting me know about the way to change the diirt.
I did not change anything. I am using your distribution on both Mac and Linux.
My linux is even a virtual machine on my Mac.

The linux dist CS-S works, but not the Mac one.

**Comment by berryma4** [comment]
you are probably need to configure the diirt dirtory on the mac.

**Comment by shengb** [comment]
Where is the diirt directory? Any instruction?

**Comment by berryma4** [comment]
in cs-studio/configuration/diirt

**Comment by shengb** [comment]
I found one:  cs-studio/configuration/diirt/datasources
there is no setting for EPICS_CA_ADDR_LIST on either Mac or Linux

**Comment by berryma4** [comment]
mac:
cs-studio/configuration/diirt/datasources/ca

the linux debian package:
 uses the one in /etc/cs-studio/diirt.

**Comment by shengb** [comment]
OK, the ca directory does not exist on Mac.
I copied it, and it now works.

@jbobnar back to our original topic:
Why the stored value is in red as shown in the picture?
<img width="953" alt="screen shot 2016-02-17 at 10 22 32 am" src="https://cloud.githubusercontent.com/assets/3741779/13114269/d51e2ef2-d560-11e5-88b2-9ddaed17efa8.png">

**Comment by shengb** [comment]
@jbobnar If the column named "Stored Value", why not "Live Setpoint" => "Live Value" as shown in above?

**Comment by berryma4** [comment]
I believe this is to distinguish between sets and reads

**Comment by shengb** [comment]
from the UI itself, how can I tell the pv listed in the 2nd column is a set point or a read back?

**Comment by jbobnar** [comment]
@shengb What you see in your screenshot has already been fixed in the version that I pushed yesterday. There was an error in the value comparison routines.

> from the UI itself, how can I tell the pv listed in the 2nd column is a set point or a read back?

That is a setpoint by definition. Whether that is an actual device setpoint or readback is beyond the knowledge of the application. The users creating the sets should pay attention to what names are used in the set file.

Since I am not the primary user of the application, I can only suggest how to name the columns. In the screenshot below I attached my proposal for all column names. Note that these are full blown tables showing all possible columns (all readback stuff can be turned on or off). The top table is for a single snapshot, the bottom one is for snapshots comparison.
 
Let me know how to name each column and I will update it accordingly. 

![screenshot1](https://cloud.githubusercontent.com/assets/5602519/13139628/c3e51bfa-d62d-11e5-8d01-4394fdb9be72.png)

**Comment by shengb** [comment]
There would be a long discussion to have set point and read back, and pair.
To simplify the problem, I would say the PV is a "Live Value" if the readback is not defined.

**Comment by berryma4** [comment]
Maybe we should ask the users?

**Comment by berryma4** [comment]
Maybe we should ask the users and get back to Jaka?

**Comment by shengb** [comment]
If there are 2 columns (readback & setpoint) like showed above, it makes sense to have a column to show live setpoint, and live readback.
If there is only one PV column, I would like to name it live value since "live setpoint" is very confusing in this case.

**Comment by berryma4** [comment]
Changing the name may be confusing.  We should aim to keep things as
consistent as possible.

**Comment by shengb** [comment]
Agree. But how to handle it is there is only one read back PV available, for example, I want to have a save set for all our BPM X/Y positions. I want to know the live value, should I still call it live set point?

**Comment by berryma4** [comment]
Can we move it to the other column?  This would seem logical, although I
don't know if possible.

**Comment by shengb** [comment]
have it in other column is fine. 
1) If there is only one PV column available, show one column; 
2)if there are more than one PV available, show column pairs is good. 

But I do not want to show them both 1) & 2) simultaneously.

**Comment by jbobnar** [comment]
@berryma4 That is not possible, because it would require the application to know if a particular PV is a readback or not. 

However, even if the application could get that information somehow, one would end up with a table, which would look very similar to the table that you see now (when you turn off all readback stuff). The only difference would be in the headers which would say Readback instead of Setpoint. Eventually that is almost identical to Guobao's suggestion of renaming the header.

**Comment by shengb** [comment]
Any progress here?
